exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,618 @@
+
+      auto publish_packet_received = co_await mqtt_client_->recv(
+          async_mqtt::filter::match, { async_mqtt::control_packet_type::publish }, asio::use_awaitable);
+
+      auto publish_packet = publish_packet_received.template get<async_mqtt::v5::publish_packet>();
+
+      logger_.trace("Received PUBLISH packet. Parsing payload...");
+
+      if (publish_packet.payload().size() == 0) {
+        logger_.warn("Received invalid payload. Cannot process packet.");
+        co_return;
+      }
+
+      auto data = publish_packet.payload()[0];
+
+      Payload payload;
+
+      bool payload_valid = payload.ParseFromArray(data.data(), data.size());
+
+      if (!payload_valid) {
+        logger_.warn("Received invalid payload. Continuing to next iteration...");
+        co_return;
+      }
+
+      logger_.trace("Payload parsed successfully. Processing payload...");
+
+      auto metric = payload.metrics(0);
+
+      if (payload.has_timestamp() && !payload.has_seq() &&
+          (publish_packet.opts().get_qos() == async_mqtt::qos::at_most_once) &&
+          (publish_packet.opts().get_retain() == async_mqtt::pub::retain::no) && (metric.name() == rebirth_metric)) {
+        logger_.trace("Conditions met. Sending NBIRTH...");
+        co_await asio::co_spawn(mqtt_client_->strand(), send_nbirth(), asio::use_awaitable);
+      } else {
+        logger_.trace(
+            "Conditions not met, payload has timestamp (should be true): {}, "
+            "has seq (should be false): {}, qos (should be at_most_once): {}, "
+            "retain (should be no): {}, "
+            "metric name (should be 'Node Control/Rebirth'): {}",
+            payload.has_timestamp(), payload.has_seq(), async_mqtt::qos_to_str(publish_packet.opts().get_qos()),
+            async_mqtt::pub::retain_to_str(publish_packet.opts().get_retain()), metric.name());
+      }
+    } catch (std::exception const& exc) {
+      logger_.error("Error in NCMD listener: {}", exc.what());
+      throw;
+    }
+  }
+
+  auto subscribe_to_ncmd_topic() -> asio::awaitable<void> {
+    logger_.trace("Starting NCMD listener...");
+
+    std::string ncmd_topic = topic_formatter(
+        { namespace_element, config_.value()._sparkplug_b_group_id, ncmd, config_.value()._sparkplug_b_edge_node_id });
+
+    auto p_id = mqtt_client_->acquire_unique_packet_id();
+
+    logger_.trace("Preparing subscription packet for topic: {}", ncmd_topic);
+
+    auto subscribe_packet = async_mqtt::v5::subscribe_packet{
+      *p_id, { { async_mqtt::allocate_buffer(ncmd_topic), async_mqtt::qos::at_most_once | async_mqtt::sub::nl::yes } }
+    };
+
+    logger_.trace("Sending subscription packet...");
+
+    co_await mqtt_client_->send(subscribe_packet, asio::use_awaitable);
+
+    logger_.trace("Subscription packet sent. Waiting for SUBACK...");
+
+    co_await mqtt_client_->recv(async_mqtt::filter::match, { async_mqtt::control_packet_type::suback }, asio::use_awaitable);
```

Comment:
Does this not have a return code that needs to be checked?

Reply:
The return value from the function is a suback packet and we can check its reason code:

https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901174

The only question is what should be done with the information in the packet, should the program be killed? If the function is returned then the caller will think that the subscribe was successful and will wait for a packet.

```cpp
auto suback_received = co_await mqtt_client_->recv(async_mqtt::filter::match, { async_mqtt::control_packet_type::suback }, asio::use_awaitable);

auto suback_packet = suback_received.template get<async_mqtt::v5::suback_packet>();

for (auto const& entry : suback_packet.entries()) {
    if (entry != async_mqtt::suback_reason_code::granted_qos_0) {
    logger_.error("Error subscribing to topic: {}, reason code: {}", ncmd_topic, async_mqtt::suback_reason_code_to_str(entry));
    std::exit(-1);
    }
}
```

Reply:
Changing the function to return an error code so that the callee the reconnects to the broker and the function is called again.